test: MiniWallet: support skipping mempool checks (feature_fee_estimation.py performance fix)

theStack opened the pull request:
MiniWallet's core method for creating txs (`create_self_transfer`) right now always executes the `testmempoolaccept` RPC to check for mempool validity or invalidity. In some test cases where we use MiniWallet to create a huge number of transactions this can lead to performance issues, in particular feature_fee_estimation.py where the execution time after MiniWallet usage (PR #24817) doubled, see https://github.com/bitcoin/bitcoin/issues/24828#issuecomment-1100058100, https://github.com/bitcoin/bitcoin/issues/24828#issuecomment-1100301980. This PR mitigates this by skipping the mempool check if the parameter `mempool_valid` is set to `False`.

As a preparatory commit, the test feature_csv_activation.py has to be adapted w.r.t. to rehashing of transactions, as we now hash all transactions immediately in `create_self_transfer` in order to get the txid (before we relied on the result of `testmempoolaccept`).

On my machine, this decreases the execution time quite noticably:

master branch:
```
$ time ./test/functional/feature_fee_estimation.py
real    3m20.771s
user    2m52.360s
sys     0m39.340s
```

PR branch:
```
$ time ./test/functional/feature_fee_estimation.py
real    2m1.386s
user    1m42.510s
sys     0m22.980s
```

Partly fixes #24828 (hopefully).

theStack committed:
test: MiniWallet: skip mempool check if `mempool_valid=False`

MiniWallet's core method for creating txs (`create_self_transfer`)
right now always executes the `testmempoolaccept` RPC to check for
mempool validity or invalidity. In some test cases where we use
MiniWallet to create a huge number of transactions this can lead
to performance issues (e.g. feature_fee_estimation.py where the
execution time after MiniWallet usage almost doubled). Providing
the possibility to skip the mempool checks is a mitigation for
this.

master branch:
$ time ./test/functional/feature_fee_estimation.py
real    3m20.771s
user    2m52.360s
sys     0m39.340s

PR branch:
$ time ./test/functional/feature_fee_estimation.py
real    2m1.386s
user    1m42.510s
sys     0m22.980s

MarcoFalke commented on test/functional/test_framework/wallet.py:
```diff
@@ -259,13 +260,17 @@ def create_self_transfer(self, *, fee_rate=Decimal("0.003"), from_node=None, utx
+        txid = tx.rehash()
+        wtxid = tx.getwtxid()
+
+        if mempool_valid is not None:
```
I don't think we need the miniwallet to check that `mempool_valid=False` is equal to tmpa["allowed"].
```suggestion
        if mempool_valid:
```

theStack replied:
Good idea, done.

theStack committed:
test: MiniWallet: always rehash after signing (P2PK mode)

Also explicitly rehash in the cases where we modify a tx after signing
in feature_csv_activation.py. Parts of this test relied on the fact that
rehashing of transactions is done in the course of calculating a block's
merkle root (`calc_merkle_root`), which only works if no hash was
calculated before due to a caching mechanism.

In the following commit the txid in MiniWallet is calculated via
`rehash()`, i.e. this doesn't work anymore and we always have to
explicitely have the right hash before we calculate the merkle root.

danielabrozzoni reviewed (approved):
tACK a498acce4514d83d8dafcebaad522a89b6dc70fa

I also experienced faster running time:

On master:
```
time ./test/functional/feature_fee_estimation.py
real    1m49.060s
user    1m10.307s
sys     0m4.977s
```

On a498acce4514d83d8dafcebaad522a89b6dc70fa:
```
time ./test/functional/feature_fee_estimation.py
real    1m27.915s
user    1m4.089s
sys     0m4.600s
```
